gh-106033: [docs] Improve C API GetItem & HasAttr notes.

gpshead opened the pull request:
This should format the notes and call out the preferred APIs to use in a more visible manner.

<!-- gh-issue-number: gh-106033 -->
* Issue: gh-106033
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
:books: Documentation preview :books:: https://cpython-previews--106047.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

gpshead committed:
[docs] Improve C API GetItem & HasAttr notes.

This should format the notes and call out the preferred APIs to use
in a more visible manner.

CAM-Gerlach commented on Doc/c-api/dict.rst:
```diff
@@ -98,9 +98,11 @@ Dictionary Objects
-   Note that exceptions which occur while calling :meth:`__hash__` and
-   :meth:`__eq__` methods will get suppressed.
-   To get error reporting use :c:func:`PyDict_GetItemWithError()` instead.
+   .. note::
```
Since this is providing a caution/warning of a problem with the function's current behavior (and recommending an alternative), I would consider a more appropriate type of admonition here, such as 

```suggestion
   .. caution::
```

or potentially even

```suggestion
   .. warning::
```

CAM-Gerlach commented on Doc/c-api/dict.rst:
```diff
@@ -98,9 +98,11 @@ Dictionary Objects
-   To get error reporting use :c:func:`PyDict_GetItemWithError()` instead.
+   .. note::
+
+      Exceptions that occur while this calls :meth:`__hash__` and
```
```suggestion
      Exceptions that occur while this calls :meth:`~object.__hash__` and
```

Fix broken reference target

CAM-Gerlach commented on Doc/c-api/dict.rst:
```diff
@@ -98,9 +98,11 @@ Dictionary Objects
+   .. note::
+
+      Exceptions that occur while this calls :meth:`__hash__` and
+      :meth:`__eq__` methods are silently ignored.
```
```suggestion
      :meth:`~object.__eq__` methods are silently ignored.
```

Likewise

CAM-Gerlach commented on Doc/c-api/dict.rst:
```diff
@@ -120,10 +122,13 @@ Dictionary Objects
-   :meth:`__eq__` methods and creating a temporary string object
-   will get suppressed.
-   To get error reporting use :c:func:`PyDict_GetItemWithError()` instead.
+   .. note::
```
Same here:

```suggestion
   .. caution::
```

or

```suggestion
   .. warning::
```

CAM-Gerlach commented on Doc/c-api/dict.rst:
```diff
@@ -120,10 +122,13 @@ Dictionary Objects
-   To get error reporting use :c:func:`PyDict_GetItemWithError()` instead.
+   .. note::
+
+      Exceptions that occur while this calls :meth:`__hash__` and
```
```suggestion
      Exceptions that occur while this calls :meth:`~object.__hash__` and
```

CAM-Gerlach commented on Doc/c-api/dict.rst:
```diff
@@ -120,10 +122,13 @@ Dictionary Objects
+   .. note::
+
+      Exceptions that occur while this calls :meth:`__hash__` and
+      :meth:`__eq__` methods or while creating the temporary ``str``
```
```suggestion
      :meth:`~object.__eq__` methods or while creating the temporary :class:`str`
```

Also, markup appropriately

CAM-Gerlach commented on Doc/c-api/object.rst:
```diff
@@ -33,9 +33,11 @@ Object Protocol
-   To get error reporting use :c:func:`PyObject_GetAttr()` instead.
+   .. note::
+
+      Exceptions that occur when this calls :meth:`__getattr__` and
```
```suggestion
      Exceptions that occur when this calls :meth:`~object.__getattr__` and
```

Likewise, fix these broken added xrefs

CAM-Gerlach commented on Doc/c-api/object.rst:
```diff
@@ -33,9 +33,11 @@ Object Protocol
+   .. note::
+
+      Exceptions that occur when this calls :meth:`__getattr__` and
+      :meth:`__getattribute__` methods are silently ignored.
```
```suggestion
      :meth:`~object.__getattribute__` methods are silently ignored.
```

CAM-Gerlach commented on Doc/c-api/object.rst:
```diff
@@ -33,9 +33,11 @@ Object Protocol
-   Note that exceptions which occur while calling :meth:`__getattr__` and
-   :meth:`__getattribute__` methods will get suppressed.
-   To get error reporting use :c:func:`PyObject_GetAttr()` instead.
+   .. note::
```
Same here:

```suggestion
   .. caution::
```

or

```suggestion
   .. warning::
```

CAM-Gerlach commented on Doc/c-api/object.rst:
```diff
@@ -44,10 +46,12 @@ Object Protocol
-   :meth:`__getattribute__` methods and creating a temporary string object
-   will get suppressed.
-   To get error reporting use :c:func:`PyObject_GetAttrString()` instead.
+   .. note::
```
And here:

```suggestion
   .. caution::
```

or

```suggestion
   .. warning::
```

CAM-Gerlach commented on Doc/c-api/object.rst:
```diff
@@ -44,10 +46,12 @@ Object Protocol
-   To get error reporting use :c:func:`PyObject_GetAttrString()` instead.
+   .. note::
+
+      Exceptions that occur when this calls :meth:`__getattr__` and
```
```suggestion
      Exceptions that occur when this calls :meth:`~object.__getattr__` and
```

CAM-Gerlach commented on Doc/c-api/object.rst:
```diff
@@ -44,10 +46,12 @@ Object Protocol
+   .. note::
+
+      Exceptions that occur when this calls :meth:`__getattr__` and
+      :meth:`__getattribute__` methods or while creating the temporary ``str``
```
```suggestion
      :meth:`~object.__getattribute__` methods or while creating the temporary :class:`str`
```

CAM-Gerlach reviewed (commented):
**Standard reminder: You can directly apply all the suggestions you want in one go by going to [`Files changed`](URL) -> Clicking <kbd>Add to batch</kbd> on each suggestion -> When done, clicking <kbd>Commit</kbd>**

Two sets of suggestions:

* Given this is cautioning/warning against problematic behavior, I suggest using the appropriate admonition type accordingly.
* Fix a number of broken references (to generic dunder methods) introduced in this PR.

Other than that, LGTM, thanks! It is definitely an improvement, as you say.

gpshead committed:
~object.

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

gpshead committed:
more ~object.

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

gpshead committed:
caution rather than note

ironically stands out less than note, but the bold caution: is still sufficient and better than what we had.  warning would stand out too much.

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

gpshead committed:
caution rather than note

ironically stands out less than note, but the bold caution: is still sufficient and better than what we had.  warning would stand out too much.

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

gpshead committed:
~object

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

gpshead committed:
~object

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

gpshead committed:
~object

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

gpshead committed:
Update Doc/c-api/dict.rst

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

gpshead committed:
Update Doc/c-api/dict.rst

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

gpshead committed:
Update Doc/c-api/dict.rst

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

gpshead committed:
Update Doc/c-api/dict.rst

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

gpshead committed:
Update Doc/c-api/object.rst

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

CAM-Gerlach reviewed (commented):
Would be LGTM, but it seems the Python docs theme is entirely missing appropriate styles for `caution` admonitions—they are not only lacking an appropriate color, but are entirely missing the whole callout box around them, too.

Therefore, until that is fixed, you'll have to either go back to using `note`, or step it up to `warning` (as is used elsewhere).

gpshead committed:
back to note

gpshead committed:
back to note
